subsys/mgmt/mcumgr/grp/settings_mgmt/src/settings_mgmt.c

```diff
@@ -0,0 +1,535 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/settings/settings.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/mgmt/handlers.h>
+#include <zephyr/mgmt/mcumgr/grp/settings_mgmt/settings_mgmt.h>
+#include <mgmt/mcumgr/util/zcbor_bulk.h>
+#include <assert.h>
+#include <limits.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#if defined(CONFIG_MCUMGR_MGMT_NOTIFICATION_HOOKS)
+#include <zephyr/mgmt/mcumgr/mgmt/callbacks.h>
+#endif
+
+#define LOG_LEVEL CONFIG_MCUMGR_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(settings_mgmt);
+
+/**
+ * Command handler: settings read
+ */
+static int settings_mgmt_read(struct smp_streamer *ctxt)
+{
+	int rc;
+	zcbor_state_t *zse = ctxt->writer->zs;
+	zcbor_state_t *zsd = ctxt->reader->zs;
+	bool ok;
+	size_t decoded;
+	struct zcbor_string key = { 0 };
+	uint32_t max_size = CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN;
+	bool limited_size = false;
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	char *key_name = NULL;
+	uint8_t *data = NULL;
+#else
+	char key_name[CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN] = { 0 };
+	uint8_t data[CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN] = { 0 };
+#endif
+
+	struct zcbor_map_decode_key_val settings_read_decode[] = {
+		ZCBOR_MAP_DECODE_KEY_DECODER("name", zcbor_tstr_decode, &key),
+		ZCBOR_MAP_DECODE_KEY_DECODER("max_size", zcbor_uint32_decode, &max_size),
+	};
+
+#if defined(CONFIG_MCUMGR_GRP_SETTINGS_ACCESS_HOOK)
+	struct settings_mgmt_access settings_access_data = {
+		.access = SETTINGS_ACCESS_READ,
+		.name = key_name,
+	};
+
+	enum mgmt_cb_return status;
+	int32_t ret_rc;
+	uint16_t ret_group;
+#endif
+
+	ok = zcbor_map_decode_bulk(zsd, settings_read_decode, ARRAY_SIZE(settings_read_decode),
+				   &decoded) == 0;
+
+	if (!ok || key.len == 0) {
+		return MGMT_ERR_EINVAL;
+	}
+
+	/* Check if the length of the user supplied key is too large for the configuration
+	 * to allow and return an error if so
+	 */
+	if (key.len >= CONFIG_MCUMGR_GRP_SETTINGS_NAME_LEN) {
+		ok = smp_add_cmd_ret(zse, MGMT_GROUP_ID_SETTINGS,
+				     SETTINGS_MGMT_RET_RC_KEY_TOO_LONG);
+		goto end;
+	}
+
+	if (max_size > CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN) {
+		max_size = CONFIG_MCUMGR_GRP_SETTINGS_VALUE_LEN;
+		limited_size = true;
+	}
+
+#ifdef CONFIG_MCUMGR_GRP_SETTINGS_BUFFER_TYPE_HEAP
+	key_name = (char *)malloc(key.len + 1);
+	data = (uint8_t *)malloc(max_size);
```

Comment:
Macros to me look worse than ifdefs for function code, they are harder to follow because then you always have to look at the macro code to see what is happening regardless if you have the option enabled or not, and can't use breakpoints